File: java/844.java
import java.util.Objects;

class Solution {

    private String parseStringForBackspace(String s) {
        while(s.contains("#")){
            int backspaceIndex = s.indexOf("#");

            if(backspaceIndex == 0) {
                s = s.substring(1);
            } else {
                s = s.replaceFirst(".[#]", "");
            }
        }

        return s;
    }

    public boolean backspaceCompare(String s, String t) {
        String newS = parseStringForBackspace(s);
        String newT = parseStringForBackspace(t);

        return Objects.equals(parseStringForBackspace(s), parseStringForBackspace(t));
    }
}

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();

        String input1 = "ab#c";
        String input2 = "ad#c";
        boolean result1 = solution.backspaceCompare(input1, input2);
        System.out.println(result1);

        String input3 = "ab##";
        String input4 = "c#d#";
        boolean result2 = solution.backspaceCompare(input3, input4);
        System.out.println(result2);

        String input5 = "a#c";
        String input6 = "b";
        boolean result3 = solution.backspaceCompare(input5, input6);
        System.out.println(result3);
    }
}

File: kotlin/844.kt
class Solution {

    private fun formatInputString(s: String): String {
        var stringToProcess = s;

        while ( stringToProcess.contains("#") ) {
            stringToProcess = when ( stringToProcess.first() == '#' ) {
                true -> stringToProcess.removePrefix("#")
                false -> {
                    val removeIndex = stringToProcess.indexOfFirst { it == '#' }
                    stringToProcess.removeRange(removeIndex - 1, removeIndex + 1)
                }
            }
        }

        return stringToProcess
    }

    fun backspaceCompare(s: String, t: String): Boolean {
        return formatInputString( s ) == formatInputString( t )
    }
}

fun main() {
    val solution = Solution()

    val input1 = "ab#c"
    val input2 = "ad#c"

    println( solution.backspaceCompare(input1, input2) )

    val input3 = "ab##"
    val input4 = "c#d#"

    println( solution.backspaceCompare(input3, input4))

    val input5 = "a#c"
    val input6 = "b"

    println( solution.backspaceCompare(input5, input6))

    val input7 = "a##c"
    val input8 = "#########"

    println( solution.backspaceCompare(input7, input8))

}